### ISSUE TITLE
Salt's presence detection

### ISSUE THREAD

**Issue by ghostsquad** [saltstack/salt#22386]
http://docs.saltstack.com/en/latest/ref/runners/all/salt.runners.manage.html#salt.runners.manage.alived

I couldn't find any documentation on what "presence detection" actually does. Can this be explained?

**Comment by rallytime** [comment]
Thanks for the request/question @ghostsquad. There are some docs [here](http://docs.saltstack.com/en/latest/topics/event/master_events.html#presence-events) that describe presence events. Does that help answer you question?

**Comment by ghostsquad** [comment]
That's an extremely brief explanation. 

1. I think the presence events doc should be linked from the runners.manage documentation.
2. So, does that mean that runners.manage.alived are only considering minions that are not in the "lost" state.
3. With default settings, `presence_events: False`, what is expected from runners.managed.alived?

**Comment by rallytime** [comment]
Yes, I agree that the explantation could use some more love. I am not particularly familiar with presence events and will default to either @cachedout or @basepi to help answer your questions more completely.

Either way, we should get some more information in the docs, for sure.

**Comment by brad-alexander** [comment]
bump https://docs.saltstack.com/en/latest/ref/runners/all/salt.runners.manage.html#module-salt.runners.manage also appears to have docstrings copied/pasted between functions and doesn't accurately differentiate the purpose of each function. These docs need some <3

**Comment by rallytime** [comment]
Thanks for the bump, @brad-alexander. I've applied some more labels here to hopefully get some more eyes on this.

**Comment by flowstate** [comment]
+1 can we get some more info here?

**Comment by rallytime** [comment]
Hi @ghostsquad, @brad-alexander, and @flowstate (thanks for the bump) I have looked at the code for `presence_events` on the master and I think the answers to the [questions above](https://github.com/saltstack/salt/issues/22386#issuecomment-90225951) are as follows:

> There should be a link to the `presence events` doc in the `runners.manage` documentation.

Agreed. This is simple to add.

> My interpretation of the documentation leads me to believe that `runners.manage.alived` will return only minions that are not in the `lost` state. Is this correct?

First, let's back up and look at what presence detection on the master is actually doing when we set `presence_events: True`. It looks to me like there are actually two presence events fired, depending on the state of the minions. When the [handle_presence](https://github.com/saltstack/salt/blob/3057f804489677e4f6d3cf96464dcf4d6661d1b1/salt/master.py#L347) function is called, we will _always_ fire the event that reports the list of currently connected minions with the `present` key in the event data. **If** there are minions that are `lost` or `new`, they will get their own event fired, which results in a `new` or `lost` key in the data sent back by the event. These events look like this in the logs:
```
[DEBUG   ] Sending event - data = {'new': ['test-minion1'], '_stamp': '2016-08-31T23:24:06.841639', 'lost': []}
[DEBUG   ] Sending event - data = {'_stamp': '2016-08-31T23:24:06.854819', 'present': ['test-minion1']}
...
[DEBUG   ] Gathering reactors for tag salt/presence/change
...
[DEBUG   ] Gathering reactors for tag salt/presence/present
...
[DEBUG   ] Sending event - data = {'new': [test-minion2'], '_stamp': '2016-08-31T23:25:11.680873', 'lost': []}
[DEBUG   ] Sending event - data = {'_stamp': '2016-08-31T23:25:11.682804', 'present': ['test-minion1', 'test-minion2']}
...
```
Now, after the looking at the code for the `salt-run manage.alived` function, it looks like the `alived` function is _only_ going to return minions that are currently connected. It does not list lost or new minions. I think the presence functionality from a runner was updated for the RAET transport, but for the other transports, we're only checking for our currently connected minions.

> With default settings, `presence_events: False`, what is expected from `runners.managed.alived`?

That is a good question. The docs do indicate that you need `presence_events` set to `True`, for this function to work. However, that doesn't appear that is the case. I set `presence_events: False` in my master config and restarted my master. While I do not see the presence events firing as mentioned above in my debug logs, I still get the list of minions that are currently connected to my salt-master:
```
root@rallytime:~# cat /etc/salt/master
# /etc/salt/master
presence_events: False
root@rallytime:~# salt-run manage.alived
- test-minion1
- test-minion2
root@rallytime:~# salt-key
Accepted Keys:
test-minion1
test-minion2
Denied Keys:
Unaccepted Keys:
Rejected Keys:
```

This result makes sense with the code, however it does seem like it might be confusing according to the docs. Rather than check for the `presence_events` setting and changing the behavior of the runner function, I am more inclined to just fix the docs on that particular issue.

I am happy to submit a pull request to update the docs to explain all of this. Before doing so, does this make sense? Am I missing anything?

**Comment by brad-alexander** [comment]
The only other thing I would add is that there are a lot of functions with the same docstring, so if you don't inherently know what **reaped**, **joined**, **allowed**, etc mean in context it doesn't really do much to explain the functions.

Specifically these functions:

> '''Print a list of all minions that are up according to Salt's presence detection (no commands will be sent to minions)'''
  
- salt.runners.manage.allowed(subset=None, show_ipv4=False)
- salt.runners.manage.alived(subset=None, show_ipv4=False)
- salt.runners.manage.joined(subset=None, show_ipv4=False)
- salt.runners.manage.list_state(subset=None, show_ipv4=False, state=None)
- salt.runners.manage.present(subset=None, show_ipv4=False)
- salt.runners.manage.reaped(subset=None, show_ipv4=False)

>'''Print a list of all minions that are NOT up according to Salt's presence detection (no commands will be sent to minions)'''

- salt.runners.manage.list_not_state(subset=None, show_ipv4=False, state=None)
- salt.runners.manage.not_alived(subset=None, show_ipv4=False)
- salt.runners.manage.not_allowed(subset=None, show_ipv4=False)
- salt.runners.manage.not_joined(subset=None, show_ipv4=False)
- salt.runners.manage.not_present(subset=None, show_ipv4=False)
- salt.runners.manage.not_reaped(subset=None, show_ipv4=False)

**Comment by flowstate** [comment]
Need to read and digest, but … thank you! This is (to my mind) the single
largest black box in saltstack. I will provide further feedback shortly.

**Comment by grinapo** [comment]
Does it even work under zmq? I switched  presence_events: True in master and the event contains "present":[]
I have tried restarting some minions, no change, zero in lists, except up and down.

**Comment by rallytime** [comment]
@grinapo Can you please post what version of salt you're using? And are the minions up and responding to the master?

**Comment by grinapo** [comment]
Yes, all the minions are just fine, and they are all present in the "up" list. 
(As a sidenote: is there a way to list which IP addresses do the minions use to communicate with the master?)

Salt Version:
           Salt: 2016.3.2
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.4.2
          gitdb: 0.6.4
      gitpython: 2.0.5
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.4
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.12+ (default, Sep  1 2016, 20:27:38)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.4.1
            ZMQ: 4.1.5
 
System Versions:
           dist: debian stretch/sid 
        machine: x86_64
        release: 4.4.16-1-pve
         system: Linux
        version: debian stretch/sid

**Comment by jimmykarily** [comment]
I'm also getting inconsistent results from manage.present and manage.up commands. I have one minions connected to the master:

```
~ $ salt-run manage.up
- 4c0669a7dcfa
```

I have presence events enabled in /etc/salt/master:

```
presence_events: True
```

I can see the presence event (which shows no present minions):

```
~ $ salt-run state.event pretty=True
salt/presence/present	{
    "_stamp": "2017-01-02T14:18:21.309058", 
    "present": []
}
```

and ``` salt-run manage.present``` this is my manage.presence output:

**Comment by jimmykarily** [comment]
I just tried setting up a minion on the host (no docker container) and it works. I get a presence/change event:

```
salt/presence/change	{
    "_stamp": "2017-01-02T14:36:30.094589", 
    "lost": [], 
    "new": [
        "Lomasi"
    ]
}
salt/presence/present	{
    "_stamp": "2017-01-02T14:36:30.095259", 
    "present": [
        "Lomasi"
    ]
}
salt/run/20170102163641407387/new	{
    "_stamp": "2017-01-02T14:36:41.721758", 
    "fun": "runner.manage.present", 
    "fun_args": [], 
    "jid": "20170102163641407387", 
    "user": "sudo_dimitris"
}
salt/run/20170102163641407387/ret	{
    "_stamp": "2017-01-02T14:36:41.801643", 
    "fun": "runner.manage.present", 
    "fun_args": [], 
    "jid": "20170102163641407387", 
    "return": [
        "Lomasi"
    ], 
    "success": true, 
    "user": "sudo_dimitris"
}
salt/presence/present	{
    "_stamp": "2017-01-02T14:37:30.545285", 
    "present": [
        "Lomasi"
    ]
}
```
and 

```
~ $ sudo salt-run manage.present
- Lomasi
```

the versions on the working minion:

```
[dimitris@Lomasi ~]$ sudo salt-minion --versions
Salt Version:
           Salt: 2016.11.1
 
Dependency Versions:
           cffi: 1.9.1
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.24.0
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.17
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.13 (default, Dec 21 2016, 07:16:46)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.4.2
            ZMQ: 4.2.0
 
System Versions:
           dist:   
        machine: x86_64
        release: 4.8.13-1-ARCH
         system: Linux
        version: Not Installed
```

**Comment by atroche** [comment]
FWIW I'm have the same problem as mentioned above, where none of my 50+ minions are showing as present (despite being there in the results of `manage.up`).

```
❯ sudo salt-minion --versions
Salt Version:
          Salt: 3003.2

Dependency Versions:
          cffi: Not Installed
      cherrypy: Not Installed
      dateutil: 2.6.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 2.10
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: 1.0.7
       msgpack: 0.5.6
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: Not Installed
      pycrypto: 2.6.1
  pycryptodome: 3.4.7
        pygit2: Not Installed
        Python: 3.6.9 (default, Jan 26 2021, 15:33:00)
  python-gnupg: 0.4.1
        PyYAML: 3.12
         PyZMQ: 17.1.2
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.2.5

System Versions:
          dist: ubuntu 18.04 Bionic Beaver
        locale: UTF-8
       machine: x86_64
       release: 5.4.0-80-lowlatency
        system: Linux
       version: Ubuntu 18.04 Bionic Beaver
```

Any ideas?

**Comment by afletch** [comment]
Seeing very unreliable data from all of the presence detection runner functions here too, I have minions which are up (and I've tried restarting them), which can be targetted and respond fine. The minion data cache contains the correct up to date data, but are always listed in `manage.not_present`

```# salt --versions
Salt Version:
          Salt: 3002.5

Dependency Versions:
          cffi: 1.14.4
      cherrypy: unknown
      dateutil: 2.4.2
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 2.11.1
       libgit2: 0.26.8
      M2Crypto: 0.35.2
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: 1.3.12
     pycparser: 2.14
      pycrypto: 2.6.1
  pycryptodome: Not Installed
        pygit2: 0.26.4
        Python: 3.6.8 (default, Nov 16 2020, 16:55:22)
  python-gnupg: Not Installed
        PyYAML: 3.13
         PyZMQ: 17.0.0
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.1.4

System Versions:
          dist: centos 7 Core
        locale: UTF-8
       machine: x86_64
       release: 3.10.0-1127.18.2.el7.x86_64
        system: Linux
       version: CentOS Linux 7 Core
```
